text/0000-overlapping_match.md

```diff
@@ -0,0 +1,183 @@
+The expression enclosed within this is only executed if none of the patterns within the
+`match/in` expression are matched. If `else` and `_` are both present then the code within the
+`else` would be marked as unreadable.
+
+Edge cases:
+1. If the `_` pattern in present in any of the contained matches and the `else` block is also
+present then a `unreachable_code` lint is emitted on the code within the `else` block
+2. Since the main reason for using a `match` is the exhaustiveness checks as long as there isn't
+an `else` block then the compiler will output an error for `non-exhaustive patterns` if not all
+branches of the `match/in` are exhaustive.
+
+Implementation Assumptions:
+1. Assuming that a `match` expression is currently implemented similar to a long chain of
+`if/else if` expressions. By this, meaning that each branch is checked one at a time and if it
+matches then it skips checking any of the other branches and jumps to the end of the expression.
+
+Implementation:
+1. This can be implemented as if it was a list of `if` expressions. And a flag to check if any
+of the branches have been visited so as to not visit the `else`
+2. To cover the `else` case the location to jump to at the end after checking all the branches
+can be stored, initially set to the start of the `else` block but if it enters any of the
+branches then it is set to immediately after the `else` block.
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+This should be called `match/in` expressions since that is the combination of keywords that are
+used similar to `for/in` expressions. This idea would be best presented as a continuation of
+existing Rust patterns since it expands on the `match` expression.
+
+This proposal should be introduced to new users right after `match` expressions are taught. This
+is the best time to teach it since it appears as an extension of that syntax and the ideas that
+are used when using `match` expressions.
+
+Within the _Rust Book_ a section after the section on the `_` placeholder could be called
+_match/in Control Flow Operator Addition_. Within this section the syntax and differences would
+be outlined. These would most notable include the multiple branches can be executed. The reader
+should be able to understand by the end of this section that this allows for multiple branches
+to be executed but it still will check for exhaustiveness when able. He should also know that
+the branches are checked top first.
+
+An example that could be used within the section:
+
+You can turn this:
+```rust
+match cmp.compare(&array[left], &array[right]) {
+    Less => {
+        merged.push(array[left]);
+        left += 1;
+    },
+    Equal => {
+        merged.push(array[left]);
+        merged.push(array[right]);
+        left += 1;
+        right += 1;
+    },
+    Greater => {
+        merged.push(array[right]);
+        right += 1;
+    }
+}
+```
+into
+```rust
+match cmp.compare(&array[left], &array[right]) in {
```

Comment:
The semantic equality of these two versions implies that `match ... in`, where a pattern is used in more than one match arm, has well-defined semantics for the order of match arms such that the statements of the first mention of pattern happens-before the statements of the second mention. This might be intuitive, or not at all - and such a divergence from the semantics of `match` should be syntactically clearer, in my opinion.

I feel this proposal is too close to the switch statements of C/C++/Java which, as far as I know, are confusing to newcomers. I think this is because intuitively, humans think of choice as exclusive - so "A or B" is parsed as "A xor B", which is why a lot of introductory books in discrete mathematics and logic has to clarify "or vs xor" (example: Logic in Computer Science, Huth & Ryan, p.4).

I think the price of repetition is worth the potential unreadability of the proposed `match ... in` construct.